File: WebApplication3/Controllers/DenunciasController.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.Entity
Imports System.Linq
Imports System.Net
Imports System.Web
Imports System.Web.Mvc
Imports WebApplication3

Namespace Controllers
    Public Class DenunciasController
        Inherits System.Web.Mvc.Controller

        Private db As New OuvidoriaDbContext

        ' GET: Denuncias
        Function Index() As ActionResult
            Return View()
        End Function

        ' GET: Denuncias/Details/5
        Function Details(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim denuncia As Denuncia = db.Denuncia.Find(id)
            If IsNothing(denuncia) Then
                Return HttpNotFound()
            End If
            Return View(denuncia)
        End Function

        ' GET: Denuncias/Create
        Function Create() As ActionResult
            Return View()
        End Function

        ' POST: Denuncias/Create
        'Para se proteger de mais ataques, habilite as propriedades específicas às quais você quer se associar. Para 
        'obter mais detalhes, veja https://go.microsoft.com/fwlink/?LinkId=317598.
        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Create(<Bind(Include:="Id,OrgaoDestino,Assunto,OrgaoAssunto,Descricao,Uf,Municipio,Local,NomeEnvolvido,FuncaoEnvolvido,OrgaoEnpresa,DataPostagem")> ByVal denuncia As Denuncia) As ActionResult
            If ModelState.IsValid Then
                db.Denuncia.Add(denuncia)
                db.SaveChanges()
                Return RedirectToAction("Index")
            End If
            Return View(denuncia)
        End Function

        ' GET: Denuncias/Edit/5
        Function Edit(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim denuncia As Denuncia = db.Denuncia.Find(id)
            If IsNothing(denuncia) Then
                Return HttpNotFound()
            End If
            Return View(denuncia)
        End Function

        ' POST: Denuncias/Edit/5
        'Para se proteger de mais ataques, habilite as propriedades específicas às quais você quer se associar. Para 
        'obter mais detalhes, veja https://go.microsoft.com/fwlink/?LinkId=317598.
        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Edit(<Bind(Include:="Id,OrgaoDestino,Assunto,OrgaoAssunto,Descricao,Uf,Municipio,Local,NomeEnvolvido,FuncaoEnvolvido,OrgaoEnpresa,DataPostagem")> ByVal denuncia As Denuncia) As ActionResult
            If ModelState.IsValid Then
                db.Entry(denuncia).State = EntityState.Modified
                db.SaveChanges()
                Return RedirectToAction("Index")
            End If
            Return View(denuncia)
        End Function

        ' GET: Denuncias/Delete/5
        Function Delete(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim denuncia As Denuncia = db.Denuncia.Find(id)
            If IsNothing(denuncia) Then
                Return HttpNotFound()
            End If
            Return View(denuncia)
        End Function

        ' POST: Denuncias/Delete/5
        <HttpPost()>
        <ActionName("Delete")>
        <ValidateAntiForgeryToken()>
        Function DeleteConfirmed(ByVal id As Integer) As ActionResult
            Dim denuncia As Denuncia = db.Denuncia.Find(id)
            db.Denuncia.Remove(denuncia)
            db.SaveChanges()
            Return RedirectToAction("Index")
        End Function

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If (disposing) Then
                db.Dispose()
            End If
            MyBase.Dispose(disposing)
        End Sub

        Function Destinatario(model As Denuncia) As ActionResult
            Return View(model)
        End Function
    End Class
End Namespace

File: WebApplication4/Controllers/DenunciasController.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Entity;
using System.Diagnostics;
using System.Linq;
using System.Net;
using System.Web;
using System.Web.Mvc;
using WebApplication4.Util;
using System.Threading.Tasks;

namespace WebApplication4.Controllers
{
    [Authorize]
    public class DenunciasController : Controller
    {
        private OuvidoriaDbContext db = new OuvidoriaDbContext();

        // GET: Denuncias/Details/5
        public ActionResult Details(Guid? id)
        {
            if (id == null)
            {
                return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
            }
            Denuncia denuncia = db.Denuncia.Find(id);
            if (denuncia == null)
            {
                return HttpNotFound();
            }
            return View(denuncia);
        }

        // GET: Denuncias/Create
        
        public ActionResult Create()
        {
            ViewBag.UsuarioId = new SelectList(db.Usuario, "Id", "Nome");
            return View();
        }

        [HttpPost]
        public async Task<ActionResult> Create(Denuncia denuncia, List<Envolvido> ListaEnvolvidos)
        {
           
            if (ModelState.IsValid)
            {
                

                // Capturando o ID do usuário
                string usuarioAutenticado = User.Identity.Name;
                Usuario usuario = db.Usuario.FirstOrDefault(x => x.Email == usuarioAutenticado);
                denuncia.UsuarioId = usuario.Id;

                foreach(Envolvido item in ListaEnvolvidos)
                {
                    item.DenunciaId = denuncia.Id;
                    db.Envolvido.Add(item);
                }

                db.Denuncia.Add(denuncia);
                
                EnvioEmail envio = new EnvioEmail();
                await Task.Factory.StartNew(() => envio.sendEmail(denuncia.Usuario, denuncia));
                
                db.SaveChanges();
                return RedirectToAction("Index", "Home");
            }

            return View(denuncia);
        }

        // GET: Denuncias/Edit/5
        public ActionResult Edit(Guid? id)
        {
            if (id == null)
            {
                return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
            }
            Denuncia denuncia = db.Denuncia.Find(id);
            if (denuncia == null)
            {
                return HttpNotFound();
            }
            ViewBag.UsuarioId = new SelectList(db.Usuario, "Id", "Nome", denuncia.UsuarioId);
            return View(denuncia);
        }

        // POST: Denuncias/Edit/5
        // Para se proteger de mais ataques, habilite as propriedades específicas às quais você quer se associar. Para 
        // obter mais detalhes, veja https://go.microsoft.com/fwlink/?LinkId=317598.
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Edit(Denuncia denuncia)
        {
            if (ModelState.IsValid)
            {
                db.Entry(denuncia).State = EntityState.Modified;
                db.SaveChanges();
                return RedirectToAction("Index", "Home");
            }
            ViewBag.UsuarioId = new SelectList(db.Usuario, "Id", "Nome", denuncia.UsuarioId);
            return View(denuncia);
        }

        // GET: Denuncias/Delete/5
        public ActionResult Delete(Guid? id)
        {
            if (id == null)
            {
                return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
            }
            Denuncia denuncia = db.Denuncia.Find(id);
            if (denuncia == null)
            {
                return HttpNotFound();
            }
            return View(denuncia);
        }

        // POST: Denuncias/Delete/5
        [HttpPost, ActionName("Delete")]
        [ValidateAntiForgeryToken]
        public ActionResult DeleteConfirmed(Guid id)
        {
            Denuncia denuncia = db.Denuncia.Find(id);
            db.Denuncia.Remove(denuncia);
            db.SaveChanges();
            return RedirectToAction("Index", "Home");
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                db.Dispose();
            }
            base.Dispose(disposing);
        }
    }
}
